components/sync15/src/migrate_state.rs

```diff
@@ -0,0 +1,199 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::record_types::MetaGlobalRecord;
+use crate::state::PersistedGlobalState;
+use crate::CollSyncIds;
+use serde_json::Value;
+
+/// Given a string persisted as our old GlobalState V1 struct, extract out
+/// the sync IDs for the collection, plus a string which should be used as the
+/// new "global persisted state" (which holds the declined engines).
+/// Returns (None, None) in early error cases (eg, invalid JSON, wrong schema
+/// version etc). Otherwise, you can expect the returned global state to be
+/// Some, even if the CollSyncIds is None (which can happen if the engine is
+/// missing, or flagged for reset)
+pub fn extract_v1_state(
```

Comment:
I'd almost prefer we kept the old global state struct decl in this file (not as `pub`) and just extracted it from that, since is kind of gross, although it's well tested so I don't really care (I also don't know how much better it would be if the decl were here).

Reply:
I did consider that, but it seemed fairly painful for not a whole lot of benefit, so I'll leave this as it is.